daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs

```diff
@@ -274,16 +285,25 @@ downloadAllReachablePackages args pid = loop Map.empty [pid]
 -- | Download the Package identified by a PackageId; fail if it doesn't exist or can't be decoded.
 downloadPackage :: LedgerArgs -> LF.PackageId -> IO LF.Package
 downloadPackage args pid = do
-  let ls :: L.LedgerService (Maybe L.Package) = do
-        lid <- L.getLedgerIdentity
-        L.getPackage lid $ convPid pid
-  runWithLedgerArgs args ls >>= \case
-    Nothing -> fail $ "Unable to download package with identity: " <> show pid
-    Just (L.Package bs) -> do
-      let mode = LFArchive.DecodeAsMain
-      case LFArchive.decodePackage mode pid bs of
-        Left err -> fail $ show err
-        Right pkg -> return pkg
+  bs <-
+    case api args of
+      Grpc -> do
+        runWithLedgerArgs args $ do
+          lid <- L.getLedgerIdentity
+          mbPkg <- L.getPackage lid $ convPid pid
+          case mbPkg of
+            Nothing -> fail $ "Unable to download package with identity: " <> show pid
```

Comment:
Calling `fail` is not the most user-friendly way to report errors. I'm absolutely aware that it was like this before but I think we can do better. Let's do that in a separate PR, if you don't mind.

Reply:
agreed.